### PR TITLE
[CAD-2671] Modify cardano-sync to enable reusing from SMASH.

### PR DESCRIPTION
**Pull request by ksaric**
https://jira.iohk.io/browse/CAD-2671

### CODE REVIEW

File: cardano-sync/src/Cardano/Sync.hs [hunk 1]
```diff
@@ -146,8 +152,15 @@ runSyncNode dataLayer plugin enp insertValidateGenesisDist =
       case genCfg of
           GenesisCardano _ bCfg _sCfg -> do
             syncEnv <- ExceptT $ mkDbSyncEnvFromConfig dataLayer (enpLedgerStateDir enp) genCfg
-            liftIO $ runDbSyncNodeNodeClient (dncPrometheusPort enc) syncEnv
-                iomgr trce plugin (cardanoCodecConfig bCfg) (enpSocketPath enp)
+            liftIO $ runDbSyncNodeNodeClient
```
Comment (erikd) [inline cardano-sync/src/Cardano/Sync.hs:155]: Why change the layout style from what was there. A good rule of thumb when submitting a PR to a project is to follow the existing coding style, eg "When in Rome do as the Romans".

File: cardano-db/src/Cardano/Db/Schema/Types.hs [hunk 2]
```diff
@@ -8,6 +8,8 @@ module Cardano.Db.Schema.Types
   , PaymentAddrHash (..)
   ) where
 
+import           Cardano.Prelude
+
```
Comment (erikd) [inline cardano-db/src/Cardano/Db/Schema/Types.hs:12]: What benefit does this actually give us?

File: cardano-db/src/Cardano/Db/Migration/Version.hs [hunk 3]
```diff
@@ -23,9 +23,9 @@ data MigrationVersion = MigrationVersion
   } deriving (Eq, Ord, Show)
 
 
-parseMigrationVersionFromFile :: String -> Maybe MigrationVersion
+parseMigrationVersionFromFile :: Text -> Maybe MigrationVersion
```
Comment (erikd) [inline cardano-db/src/Cardano/Db/Migration/Version.hs:26]: Why are these types changing from `String` to `Text`? It does not seem to be related to "Modify cardano-sync to enable reusing from SMASH:.

File: cardano-db/src/Cardano/Db/Migration/Version.hs [hunk 4]
```diff
@@ -40,18 +40,18 @@ nextMigrationVersion (MigrationVersion _stage ver _date) = do
   (y, m, d) <- Time.toGregorian . Time.utctDay <$> Time.getCurrentTime
   pure $ MigrationVersion 2 (ver + 1) (fromIntegral y * 10000 + m * 100 + d)
 
-renderMigrationVersion :: MigrationVersion -> String
+renderMigrationVersion :: MigrationVersion -> Text
 renderMigrationVersion mv =
-  List.intercalate "-"
+  toS $ intercalate "-"
```
Comment (erikd) [inline cardano-db/src/Cardano/Db/Migration/Version.hs:45]: Why are we moving from qualified ` List.intercalate` to unqualified `intercalate` ? I know `Cardano.Prelude` exposes `intercalate` but I do not see a good reason to switch from qualified to unqualified usage.

File: cardano-db/cardano-db.cabal [hunk 5]
```diff
@@ -20,6 +20,9 @@ extra-source-files:     CHANGELOG.md
 
 library
   default-language:     Haskell2010
+  default-extensions:
+    NoImplicitPrelude
+    OverloadedStrings
```
Comment (erikd) [inline cardano-db/cardano-db.cabal:25]: This seems un-related to the aim of the PR. In general I think defining default extensions in the cabal file is a bad idea. Just about the only one where I think this is not a bad idea is `NoImplicitPrelude` .

File: cardano-db/src/Cardano/Db/Query.hs [hunk 6]
```diff
@@ -65,6 +65,7 @@ module Cardano.Db.Query
   , unValueSumAda
   ) where
 
+import           Cardano.Prelude hiding (Meta, from, isJust, isNothing, maybeToEither, on)
```
Comment (erikd) [inline cardano-db/src/Cardano/Db/Query.hs:68]: This change does not offer a net benefit.

File: cardano-sync/src/Cardano/Sync.hs (hunk 1 above)
Comment (ksaric) [reply cardano-sync/src/Cardano/Sync.hs:155]: Fine, I can revert that.

File: cardano-db/src/Cardano/Db/Schema/Types.hs (hunk 2 above)
Comment (ksaric) [reply cardano-db/src/Cardano/Db/Schema/Types.hs:12]: It's a standard prelude we are supposed to use? I guess I don't have to talk about the benefits of a custom Prelude?

File: cardano-db/src/Cardano/Db/Migration/Version.hs (hunk 3 above)
Comment (ksaric) [reply cardano-db/src/Cardano/Db/Migration/Version.hs:26]: Well. 
First, I noticed that it would be better if I add `cardano-prelude` in a single module.
Then, there was a collision in that module between `cardano-prelude` and base Prelude.
Then I added the extension for the whole library which I consider worthy, `NoImplicitPrelude`.
Then other modules started complaining.
Then I fixed the other modules.

After a while, I realized it might have been better to do this in another PR, but you don't know how extensive the changes are until you do them.

File: cardano-db/src/Cardano/Db/Migration/Version.hs (hunk 4 above)
Comment (ksaric) [reply cardano-db/src/Cardano/Db/Migration/Version.hs:45]: https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

File: cardano-db/cardano-db.cabal (hunk 5 above)
Comment (ksaric) [reply cardano-db/cardano-db.cabal:25]: https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

File: cardano-db/src/Cardano/Db/Query.hs (hunk 6 above)
Comment (ksaric) [reply cardano-db/src/Cardano/Db/Query.hs:68]: https://github.com/input-output-hk/cardano-db-sync/pull/527#discussion_r575045248

File: cardano-db/cardano-db.cabal (hunk 5 above)
Comment (ksaric) [reply cardano-db/cardano-db.cabal:25]: I think that using `Text` is what you want to do, in most cases, so this is something that I used quite a lot. Additionally, the current `cardano-prelude` does not export `String`. So there is that.

File: cardano-db/src/Cardano/Db/Schema/Types.hs (hunk 2 above)
Comment (erikd) [reply cardano-db/src/Cardano/Db/Schema/Types.hs:12]: > It's a standard prelude we are supposed to use? 

No, there is no mandate.

**Review by erikd — CHANGES_REQUESTED** [review]
* Rebase against master.
* The change that includes `Cardano.Prelude` everywhere needs to be split out into a separate commit.
* Remove language pragmas from cabal files and put them in the files that need then (only).
* Fix all un-used import warnings (due ot use of `Cardano.Prelude`).

**Comment by ksaric** [comment]
I rebased against `master` and reverted most of the changes related to the `cardano-prelude` and removed the excess pragmas. Should be ok for now, we can always propagate it through the codebase in later PRs, if so required.

**Review by erikd — APPROVED** [review]
